src/core/congestion_control.c

```diff
@@ -160,59 +160,39 @@ QuicCongestionControlGetSendAllowance(
+    } else if (
+        !TimeSinceLastSendValid ||
+        !Connection->Settings.PacingEnabled ||
+        !Connection->Paths[0].GotFirstRttSample ||
+        Connection->Paths[0].SmoothedRtt < MS_TO_US(15)) {
         //
-        // Pacing is disabled or we don't have an RTT sample yet, so just send
-        // everything we can.
+        // We're not in the necessary state to pace.
         //
         SendAllowance = Cc->CongestionWindow - Cc->BytesInFlight;
 
     } else {
+
         //
-        // Try to pace: if the window and RTT are large enough, the window can
-        // be split into chunks which are spread out over the RTT.
-        // SendAllowance will be set to the size of the next chunk.
+        // We are pacing, so split the congestion window into chunks which are
+        // spread out over the RTT. Calculate the current send allowance (chunk
+        // size) as the time since the last send times the pacing rate (CWND / RTT).
         //
-        uint32_t MinChunkSize = QUIC_SEND_PACING_MIN_CHUNK * Connection->Paths[0].Mtu;
-        if (Connection->Paths[0].SmoothedRtt < MS_TO_US(QUIC_SEND_PACING_INTERVAL) ||
-            Cc->CongestionWindow < MinChunkSize ||
-            !TimeSinceLastSendValid) {
-            //
-            // Either the RTT is too small (i.e. it cannot be split into
-            // multiple intervals based on the timer granularity) or the window
-            // is too small (i.e. it cannot be split into chunks larger than
-            // MinChunkSize) for us to use pacing, or this is the first send,
-            // in which case the pacing formula (which uses the time since the
-            // last send) is invalid.
-            //
-            SendAllowance = Cc->CongestionWindow - Cc->BytesInFlight;
-
-        } else {
-
-            //
-            // We are pacing, so calculate the current chunk size based on how
-            // long it's been since we sent the previous chunk.
-            //
 
-            //
-            // Since the window grows via ACK feedback and since we defer
-            // packets when pacing, using the current window to calculate the
-            // pacing interval is not quite as aggressive as we'd like. Instead,
-            // use the predicted window of the next RTT.
-            //
-            uint64_t EstimatedWnd = QuicCongestionControlPredictNextWindow(Cc);
+        //
+        // Since the window grows via ACK feedback and since we defer packets
+        // when pacing, using the current window to calculate the pacing
+        // interval is not quite as aggressive as we'd like. Instead, use the
+        // predicted window of the next round trip.
+        //
+        uint64_t EstimatedWnd = QuicCongestionControlPredictNextWindow(Cc);
 
-            SendAllowance =
-                (uint32_t)((EstimatedWnd * TimeSinceLastSend) / Connection->Paths[0].SmoothedRtt);
-            if (SendAllowance < MinChunkSize) {
```

Comment:
My only thought/concern on this PR is that if we flush send almost immediately after getting an ACK when we were CC blocked, we might end up having a really small SendAllowance. Is that Ok? Or should we keep a 1 MTU minimum?

Reply:
Can you elaborate?

Reply:
Ok, assume at t=0 we're currently blocked on CC and we receive an ACK for inflight data. That reduces inflight below cwnd and now we're no longer blocked on CC, so we queue a "flush send" operation. We also reset the last send time to Now(). The flush operation is then processed a few microseconds later. We then calculate send allowance based on the elapsed time of Now() - LastSend, multiple that by the calculated rate. It might result in a very small value (perhaps << MTU).

Reply:
Data looks good. I will continue to test and possibly improve this area in the future. For now, I leave it as is.